# Projekt 2 - ELECTRE III i PROMETHEE

#### Filip Marciniak 148148, Szymon Pasternak 148146

# Opis zbioru danych

1. Jaka jest domena problemu? 

W problemie rozważamy koszykarzy z ligi NBA grających w sezonie 2023/2024.

2. Jakie jest źródło danych?

Dane pochodzą z oficjalnej strony NBA - https://www.nba.com/stats/players/traditional.

3. Jaki jest punkt widzenia decydenta?

Jako decydent zależy nam na znalezieniu rankingu wszystkich rozważanych koszykarzy.

4. Jaka jest liczba rozważanych wariantów? Czy w oryginalnym zbiorze danych było ich więcej?

W przygotowanym zbiorze danych jest 38 wariantów. Ze strony NBA stats wybranych zostało 40 korzykarzy, którzy mają najwięcej punktów na mecz, a następnie usunięto takich, którzy mają poniżej 30 rozegranych meczów, ponieważ mogliby zaburzać dane. W oryginalnym zbiorze dostępnych jest 549 koszykarzy.

In [1]:
import pandas as pd
import numpy as np
pd.set_option('display.float_format', lambda x: '%.5f' % x)

Poniżej przedstawiony został pełen zbiór danych.

In [2]:
df = pd.read_csv('./Basketball_players.csv', sep=';')
df

PLAYER      PTS      FG%      3P%      FT%      REB  \
0               Joel Embiid 35.30000 53.30000 36.60000 88.30000 11.30000   
1               Luka Doncic 33.90000 48.70000 37.40000 78.70000  9.10000   
2     Giannis Antetokounmpo 30.70000 61.60000 28.80000 65.70000 11.50000   
3   Shai Gilgeous-Alexander 30.40000 54.00000 36.70000 87.10000  5.60000   
4              Kevin Durant 27.60000 52.90000 42.20000 85.50000  6.80000   
5          Donovan Mitchell 27.40000 46.80000 37.30000 86.20000  5.30000   
6             Jalen Brunson 27.40000 47.70000 40.10000 84.50000  3.60000   
7              Jayson Tatum 27.30000 47.20000 37.80000 83.20000  8.20000   
8              Devin Booker 27.20000 49.20000 36.40000 88.10000  4.80000   
9             Stephen Curry 26.50000 44.70000 40.30000 92.30000  4.40000   
10             De'Aaron Fox 26.50000 46.60000 36.30000 73.00000  4.50000   
11               Trae Young 26.40000 42.60000 37.10000 85.60000  2.70000   
12          Anthony Edwards 26.20000 46.40000 37.30000 82.70000  5.50000   
13             Nikola Jokic 26.10000 58.10000 35.00000 82.80000 12.20000   
14             Tyrese Maxey 25.80000 44.90000 37.90000 86.70000  3.70000   
15             LeBron James 25.30000 53.00000 40.50000 74.80000  7.30000   
16             Kyrie Irving 25.30000 49.10000 40.70000 90.50000  5.00000   
17            Anthony Davis 24.70000 55.30000 28.90000 81.50000 12.60000   
18           Damian Lillard 24.40000 42.60000 35.80000 92.30000  4.40000   
19            Julius Randle 24.00000 47.20000 31.10000 78.10000  9.20000   
20             Desmond Bane 23.70000 46.40000 38.10000 87.00000  4.40000   
21            Kawhi Leonard 23.70000 52.40000 41.50000 88.70000  6.10000   
22            DeMar DeRozan 23.30000 47.40000 33.00000 85.90000  4.30000   
23             Jaylen Brown 23.30000 50.40000 35.40000 69.90000  5.50000   
24          Lauri Markkanen 23.20000 48.00000 39.90000 89.90000  8.20000   
25          Zion Williamson 22.80000 58.60000 35.70000 70.20000  5.80000   
26          Dejounte Murray 22.70000 46.40000 36.60000 80.40000  5.30000   
27          Anfernee Simons 22.60000 43.00000 38.50000 91.60000  3.60000   
28              Paul George 22.50000 47.00000 40.90000 89.70000  5.30000   
29           Paolo Banchero 22.50000 46.40000 35.90000 71.00000  6.80000   
30               Kyle Kuzma 22.40000 46.70000 33.90000 77.90000  6.50000   
31        Jaren Jackson Jr. 22.30000 44.20000 32.00000 80.50000  5.60000   
32          Cade Cunningham 22.30000 44.80000 35.00000 86.60000  4.30000   
33       Karl-Anthony Towns 22.10000 50.60000 42.30000 87.40000  8.40000   
34            Pascal Siakam 21.80000 53.50000 34.10000 73.00000  7.00000   
35               Cam Thomas 21.80000 44.40000 35.70000 84.90000  3.20000   
36             Jimmy Butler 21.40000 50.10000 41.50000 85.60000  5.40000   
37           Alperen Sengun 21.10000 53.70000 29.70000 69.30000  9.30000   

        AST     TOV      PF      +/-  
0   5.70000 3.70000 2.90000  6.90000  
1   9.80000 4.00000 2.10000  3.50000  
2   6.40000 3.40000 2.90000  4.60000  
3   6.30000 2.20000 2.50000  7.90000  
4   5.20000 3.30000 1.70000  3.80000  
5   6.10000 2.90000 2.20000  7.00000  
6   6.50000 2.40000 1.90000  6.10000  
7   4.90000 2.50000 2.00000  8.80000  
8   6.90000 2.50000 3.00000  4.60000  
9   5.00000 2.80000 1.60000  2.00000  
10  5.60000 2.70000 2.60000  2.60000  
11 10.80000 4.30000 2.10000 -2.20000  
12  5.20000 3.10000 1.80000  6.50000  
13  9.00000 2.90000 2.50000  8.20000  
14  6.10000 1.70000 2.20000  3.50000  
15  8.10000 3.30000 1.20000  2.70000  
16  5.30000 1.70000 1.90000  3.70000  
17  3.60000 2.00000 2.40000  1.10000  
18  7.10000 2.50000 1.80000  5.80000  
19  5.00000 3.50000 2.70000  4.70000  
20  5.50000 2.70000 2.90000 -4.80000  
21  3.60000 1.70000 1.40000  5.80000  
22  5.30000 1.50000 2.10000 -1.00000  
23  3.60000 2.30000 2.60000  6.70000  
24  2.00000 1.40000 1.80000  0.40000  
25  5.00000 2.70000 2.30000  2.30000  
26 

5. Opisz jeden z rozważanych wariantów - jak jest oceniony na poszczególnych kryteriach, jak dobry jest na tle innych.

**Nikola Jokic**:
- PTS - uzyskuje dość dobry wynik, znajduje się na 14 miejscu na tym kryterium,
- FG% - na tym kryterium jest jednym z najlepszych wariantów,
- 3P% - na tym kryterium znajduje się mniej więcej w połowie,
- FT% - podobnie jak kryterium 3P%
- REB - wyróżnia się spośród innych wariantów dwucyfrowym wynikiem na tym kryterium, są jedynie 4 oceny dwucyfrowe,
- AST - na tym kryterium również jest jednym z lepszych wariantów
- TOV - na tym kryterium, radzi sobie gorzej niż inne warianty, ale o niewielką wartość
- PF - na tym kryterium radzi sobie średnio
- +/- - ostatnie kryterium zdecydowanie wyróżnia go spośród innych wariantów, uzyskuje na nim jedną z najlepszych ocen.

6. Ile rozważasz kryteriów? Czy było ich więcej w oryginalnym zbiorze danych?
  
W problemie rozważanych jest 9 kryteriów. W oryginalnym zbiorze danych było ich więcej natomiast wybrane zostały takie, które uśredniają niektóre z dostępnych oryginalnie oraz najważniejsze, takie jak np. liczba punktów na mecz.

7. Jakie są typy poszczególnych kryteriów? Scharakteryzuj je pod kątem:

Na wszystkich z kryteriów uzyskane wartości są ciągłe, ponieważ są to statystyki średnio na mecz.
- PTS - ZYSK - średnia liczba punktów na mecz
- FG% - ZYSK - procent trafionych rzutów
- 3P% - ZYSK - procent trafionych rzutów 3 punktowych
- FT% - ZYSK - procent trafionych rzutów osobistych
- REB - ZYSK - średnia liczba zbiórek na mecz
- AST - ZYSK - średnia liczba asyst na mecz
- TOV - KOSZT - średnia liczba strat na mecz
- PF - KOSZT - średnia liczba fauli na mecz
- +/- - ZYSK - średnia różnica pomiędzy punktami zdobytymi i straconymi przez drużynę w czasie kiedy gracz był obecny na boisku na mecz

Osiągane wartości:

In [3]:
df.describe()

PTS      FG%      3P%      FT%      REB      AST      TOV       PF  \
count 38.00000 38.00000 38.00000 38.00000 38.00000 38.00000 38.00000 38.00000   
mean  25.10263 49.10263 36.68158 82.55526  6.38684  5.48947  2.61053  2.27105   
std    3.31063  4.59497  3.54476  7.19511  2.55451  1.89106  0.69895  0.56660   
min   21.10000 42.60000 28.80000 65.70000  2.70000  2.00000  1.40000  1.20000   
25%   22.52500 46.40000 35.10000 78.25000  4.42500  4.57500  2.05000  1.90000   
50%   24.20000 47.55000 36.65000 85.20000  5.55000  5.30000  2.65000  2.20000   
75%   26.50000 52.77500 39.55000 87.32500  7.97500  6.27500  3.05000  2.60000   
max   35.30000 61.60000 42.30000 92.30000 12.60000 10.80000  4.30000  3.60000   

           +/-  
count 38.00000  
mean   2.29474  
std    4.46070  
min   -7.80000  
25%   -0.05000  
50%    3.30000  
75%    5.80000  
max    8.80000

8. Jakie są wagi poszczególnych kryteriów?

Aby ustalić wagi danych kryteriów zastosowana została procedura SRF. Poniżej zaprezentowane zostało ustawienie kart:

<div>
    <img src="./criteria_cards.png" alt="cards" width="700"/>
</div>

`Ratio Z` ustawione zostało na 6. Po wykonaniu obliczeń wagi przedstawiają się nastepująco:

| CRITERION | WEIGHT |
| --------- | ------ |
| PTS | 6 |
| AST, REB | 3.86 |
| FG%, 3P% | 3.14 |
| TOV | 2.43 |
| FT%, +/- | 1.71 |
| PF | 1 |

In [4]:
CRITERIA = {
    'PTS': {'type': 'gain', 'k': 6, 'q': 4, 'p': 8, 'v': 12},
    'FG%': {'type': 'gain', 'k': 3.14, 'q': 3, 'p': 4, 'v': 8},
    '3P%': {'type': 'gain', 'k': 3.14, 'q': 4, 'p': 5, 'v': 11},
    'FT%': {'type': 'gain', 'k': 1.71, 'q': 3, 'p': 7, 'v': 15},
    'REB': {'type': 'gain', 'k': 3.86, 'q': 2, 'p': 5, 'v': 9},
    'AST': {'type': 'gain', 'k': 3.86, 'q': 1.5, 'p': 3, 'v': 8},
    'TOV': {'type': 'cost', 'k': 2.43, 'q': 1, 'p': 1.5, 'v': 2.5},
    'PF':  {'type': 'cost', 'k': 1, 'q': 1, 'p': 1.5, 'v': 2},
    '+/-': {'type': 'gain', 'k': 1.71, 'q': 3, 'p': 4, 'v': 8},
}

9. Czy istnieją jakieś dominacje w zbiorze wariantów? Jeśli tak, podaj je.

W zbiorze istnieją warianty zdominowane przez inne.

In [5]:
for alt_1_id in df.index:
    for alt_2_id in df.index:
        if alt_1_id == alt_2_id:
            continue
        
        if all([True if df.loc[alt_1_id, cr] < df.loc[alt_2_id, cr] else False for cr, config in CRITERIA.items() if config['type'] == 'gain']) \
               and all([True if df.loc[alt_1_id, cr] > df.loc[alt_2_id, cr] else False for cr, config in CRITERIA.items() if config['type'] == 'cost']):
            print(f"{df.loc[alt_2_id, 'PLAYER'].ljust(23)} dominates {df.loc[alt_1_id, 'PLAYER']}")
        

Shai Gilgeous-Alexander dominates De'Aaron Fox
Nikola Jokic            dominates Julius Randle
Shai Gilgeous-Alexander dominates Jaylen Brown
Jayson Tatum            dominates Kyle Kuzma
Kawhi Leonard           dominates Jaren Jackson Jr.
Kyrie Irving            dominates Cam Thomas
Kawhi Leonard           dominates Cam Thomas


10. Jak teoretycznie powinien prezentować się najlepszy wariant w Twoim zbiorze danych?

Wariant najlepszy powinien osiągać następujące oceny na poszczególnych kryteriach:
| PTS |	FG% | 3P% | FT% | REB | AST | TOV | PF | +/- |
| --- | --- | --- | --- | --- | --- | --- | --- | --- | 
| 35.30000 | 61.60000 | 42.30000 | 92.30000 | 12.60000 | 10.80000 | 4.30000 | 3.60000| 8.80000 |

11. Spośród rozważanych wariantów, który wydaje się być najlepszy/lepszy od innych i dlaczego?

Najlepsi będą koszykarze, którzy mają najwięcej zdobytych punktów na mecz, ponieważ to kryterium posiada największą wagę. Najlepszy więc wydaje się być Joel Embiid.

12. Spośród rozważanych wariantów, który wydaje się być najsłabszy/ znacząco słabszy od innych i dlaczego?

Najgorsi będą koszykarze, którzy mają najmniej zdobytych punktów na mecz, ponieważ to kryterium posiada największą wagę. Najgorszy więc wydaje się być Alperen Sengun.

# ELECTRE III

In [6]:
def calculate_marginal_concordance(ga, gb, q, p, type_='gain'):

    if type_ == 'gain':
        if ga - gb >= -q:
            return 1
        elif ga - gb < -p:
            return 0
        else:
            return (p-(gb-ga))/(p-q)
    elif type_ == 'cost':
        if ga - gb <= q:
            return 1
        elif ga - gb > p:
            return 0
        else:
            return (p-(ga-gb))/(p-q)
    else:
        raise Exception("wrong criterion type")
    

def calculate_marginal_discordance(ga, gb, p, v, type_='gain'):

    if type_ == 'gain':
        if ga - gb <= -v:
            return 1
        elif ga - gb >= -p:
            return 0
        else:
            return ((gb-ga)-p)/(v-p)
    elif type_ == 'cost':
        if ga - gb >= v:
            return 1
        elif ga - gb <= p:
            return 0
        else:
            return (v-(ga-gb))/(v-p)
    else:
        raise Exception("wrong criterion type")
    

## Calculating Comprehensive Concordance

In [7]:
# Comprehensive Concordance will be stored inside df_C variable which is a NxN dataframe
df_C = pd.DataFrame(1.0, index=df.index, columns=df.index)
weights = sum([config['k'] for config in CRITERIA.values()])

for alt_1_id in df_C.index:
    for alt_2_id in df_C.index:
        if alt_1_id == alt_2_id:
            continue

        comprehensive_concordance_index = 0
        # calculate marginal concordance index for each criterion and add it to the cc index with appropriate weight
        for criterion, config in CRITERIA.items():
            marginal_concordance = calculate_marginal_concordance(
                ga=df.loc[alt_1_id, criterion],
                gb=df.loc[alt_2_id, criterion],
                q=config['q'],
                p=config['p'],
                type_=config['type']
            )
            comprehensive_concordance_index += config['k'] * marginal_concordance
        # normalize the cc index
        comprehensive_concordance_index /= weights
        df_C.loc[alt_1_id, alt_2_id] = comprehensive_concordance_index

## Calculate Outranking Credibility

In [8]:
df_sigma = pd.DataFrame(1.0, index=df.index, columns=df.index)

for alt_1_id in df_C.index:
    for alt_2_id in df_C.index:
        if alt_1_id == alt_2_id:
            continue

        C = sigma = df_C.loc[alt_1_id, alt_2_id]
        # calculate marginal discordance for each criterion
        for criterion, config in CRITERIA.items():
            marginal_discordance = calculate_marginal_discordance(
                ga=df.loc[alt_1_id, criterion],
                gb=df.loc[alt_2_id, criterion],
                p=config['p'],
                v=config['v'],
                type_=config['type']
            )
            # if the marginal discordance is greater than the result of concordance test,
            # take it into account while calculating outranking credibility (sigma)
            if marginal_discordance > C:
                sigma = sigma * (1-marginal_discordance)/(1-C)
        df_sigma.loc[alt_1_id, alt_2_id] = sigma

np.fill_diagonal(df_sigma.values, -1)

## Distillation

In [9]:
def distillation(df_m, upper_threshold=None, descending=True):

    # compute upper threshold
    if upper_threshold is None:
        upper_threshold = df_m.values.max()
    
    # compute lower threshold
    s = -0.15 * upper_threshold + 0.3
    lower_threshold = upper_threshold - s
    lower_threshold = df_m[(df_m < lower_threshold) & (df_m > -1)].max().max()
    if np.isnan(lower_threshold):
        lower_threshold = 0
    
    # print(df_m)
    # print(f"{s=}, {lower_threshold=}, {type(lower_threshold)} {upper_threshold=}")

    # if upper threshold is equal to 0, then stop the algorithm
    if upper_threshold == 0:
        return df_m.index

    # save only relations that satisfy given conditions to the df_f (df filtered)
    df_f = pd.DataFrame(-1.0, index=df_m.index, columns=df_m.index)
    for alt_1_id in df_m.index:
        for alt_2_id in df_m.columns:
            if alt_1_id == alt_2_id:
                continue
            sigma = df_m.loc[alt_1_id, alt_2_id]
            if sigma > lower_threshold and sigma > df_m.loc[alt_2_id, alt_1_id] + s:
                df_f.loc[alt_1_id, alt_2_id] = sigma

    # compute the strength, weakness and quality
    df_swq = pd.DataFrame(0, index=df_m.index, columns=['strength', 'weakness', 'quality'])
    df_swq['strength'] = df_f.apply(lambda row: (row != -1.0).sum(), axis=1)
    df_swq['weakness'] = df_f.apply(lambda col: (col != -1.0).sum(), axis=0)
    df_swq['quality'] = df_swq['strength'] - df_swq['weakness']

    # take subset with maximal quality
    df_swq['rank'] = df_swq['quality'].rank(ascending=False, method='min')
    if descending:
        df_mq = df_swq[df_swq['rank'] == 1]
    else:
        df_mq = df_swq[df_swq['rank'] == df_swq['rank'].max()]
    if len(df_mq.index) == 1:
        return df_mq.index
    else:
        return distillation(
            df_m.loc[df_mq.index, df_mq.index],
            upper_threshold=lower_threshold,
            descending=descending
        )

## Adding descending preorder and ascending preorder to the main dataframe

In [10]:
df['desc_preorder'] = 0
df['asc_preorder'] = 0

descending_results = []
df_copy = df_sigma.copy()
position = 1
while len(df_copy.index) > 0:
    alternatives = distillation(df_copy, descending=True)
    df_copy = df_copy.loc[~df_copy.index.isin(alternatives), ~df_copy.index.isin(alternatives)]
    df.loc[alternatives, 'desc_preorder'] = position
    descending_results.append([df.loc[alt, "PLAYER"] for alt in alternatives])
    position += 1
print(f"{descending_results=}")

ascending_results = []
df_copy = df_sigma.copy()
position = len(df_sigma)
while len(df_copy.index) > 0:
    alternatives = distillation(df_copy, descending=False)
    df_copy = df_copy.loc[~df_copy.index.isin(alternatives), ~df_copy.index.isin(alternatives)]
    df.loc[alternatives, 'asc_preorder'] = position
    ascending_results.insert(0, [df.loc[alt, "PLAYER"] for alt in alternatives])
    position -= 1
df['asc_preorder'] = df['asc_preorder'] - df['asc_preorder'].min() + 1
print(f"{ascending_results=}")

descending_results=[['Joel Embiid'], ['Nikola Jokic'], ['Shai Gilgeous-Alexander'], ['Kevin Durant'], ['Kawhi Leonard'], ['Jayson Tatum'], ['Luka Doncic'], ['Jimmy Butler'], ['Kyrie Irving'], ['Paul George'], ['LeBron James'], ['Karl-Anthony Towns'], ['Donovan Mitchell'], ['Jalen Brunson', 'Devin Booker'], ['Tyrese Maxey', 'Damian Lillard'], ['Stephen Curry', 'Lauri Markkanen', 'Zion Williamson'], ['Anthony Edwards'], ['Anthony Davis'], ["De'Aaron Fox", 'DeMar DeRozan', 'Pascal Siakam'], ['Desmond Bane', 'Jaylen Brown'], ['Trae Young'], ['Cade Cunningham'], ['Giannis Antetokounmpo'], ['Julius Randle', 'Alperen Sengun'], ['Anfernee Simons'], ['Dejounte Murray', 'Paolo Banchero'], ['Kyle Kuzma', 'Jaren Jackson Jr.', 'Cam Thomas']]
ascending_results=[['Joel Embiid', 'Nikola Jokic'], ['Luka Doncic'], ['Shai Gilgeous-Alexander'], ['Giannis Antetokounmpo', 'Kawhi Leonard'], ['Kyrie Irving', 'Paul George'], ['Jayson Tatum'], ['Jimmy Butler'], ['Kevin Durant'], ['Jalen Brunson', 'Devin Booker'

## Creating df_final with final graph

Final preorder is the intersection of descending and ascending preorders

In [11]:
df_final = pd.DataFrame(0, index=df.index, columns=df.index)

for alt_1_id in df_final.index:
    for alt_2_id in df_final.columns:
        if alt_1_id == alt_2_id:
            continue

        alt_1_d = df.loc[alt_1_id, "desc_preorder"]
        alt_1_a = df.loc[alt_1_id, "asc_preorder"]
        alt_2_d = df.loc[alt_2_id, "desc_preorder"]
        alt_2_a = df.loc[alt_2_id, "asc_preorder"]

        if alt_1_d < alt_2_d and alt_1_a < alt_2_a:
            df_final.loc[alt_1_id, alt_2_id] = 1
        elif (alt_1_d < alt_2_d and alt_1_a == alt_2_a) or (alt_1_d == alt_2_d and alt_1_a < alt_2_a):
            df_final.loc[alt_1_id, alt_2_id] = 1
        elif alt_1_d == alt_2_d and alt_1_a == alt_2_a:
            df_final.loc[alt_1_id, alt_2_id] = 2

# ---------------------------------------------------------------------------------------------------- #
# Code used to produce output for graph visualization
            
# relations = []
# for alt_1_id in df_final.index:
#     for alt_2_id in df_final.columns:
#         if alt_1_id == alt_2_id:
#             continue
#         if df_final.loc[alt_1_id, alt_2_id]:
#             relations.append({"alternative_1": alt_1_id, "alternative_2": alt_2_id})

# alternatives = []
# for alt_1_id in df.index:
#     alternatives.append({"id": alt_1_id, "name": df.loc[alt_1_id, "PLAYER"]})

# print(relations)
# print(alternatives)

## Finding rank for each alternative

In [12]:
def find_longest_paths(df):
    def dfs(node, path):
        if node not in parents:
            return
        for parent in parents[node]:
            dfs(parent, path + [parent])
        longest_paths[node] = max(longest_paths[node], path, key=len)

    # find parents for each node
    parents = {}
    for row in df.index:
        parents[row] = [col for col in df.columns if df.loc[row, col] == 1]

    # initialize path for each node with empty list
    longest_paths = {node: [] for node in df.columns}

    # perform DFS from each node
    for node in df.columns:
        dfs(node, [])

    return longest_paths

longest_paths = find_longest_paths(df_final)
df['ranks'] = [len(path)+1 for path in longest_paths.values()]

## Finding median rank for each alternative

In [13]:
df['median'] = 0
position = 1
for rank_value in df['ranks'].sort_values().unique():

    rank_group = df[df['ranks'] == rank_value].copy()
    rank_group['plus'] = rank_group['desc_preorder'] + rank_group['asc_preorder']
    rank_group = rank_group.sort_values('plus')
    for unique_plus in rank_group['plus'].unique():
        rank_group_unique_plus = rank_group[rank_group['plus'] == unique_plus]
        for alt_id in rank_group_unique_plus.index:
            df.loc[alt_id, "median"] = position

        position += 1

df.sort_values(['ranks', 'median'])

PLAYER      PTS      FG%      3P%      FT%      REB  \
0               Joel Embiid 35.30000 53.30000 36.60000 88.30000 11.30000   
13             Nikola Jokic 26.10000 58.10000 35.00000 82.80000 12.20000   
3   Shai Gilgeous-Alexander 30.40000 54.00000 36.70000 87.10000  5.60000   
1               Luka Doncic 33.90000 48.70000 37.40000 78.70000  9.10000   
21            Kawhi Leonard 23.70000 52.40000 41.50000 88.70000  6.10000   
4              Kevin Durant 27.60000 52.90000 42.20000 85.50000  6.80000   
7              Jayson Tatum 27.30000 47.20000 37.80000 83.20000  8.20000   
16             Kyrie Irving 25.30000 49.10000 40.70000 90.50000  5.00000   
2     Giannis Antetokounmpo 30.70000 61.60000 28.80000 65.70000 11.50000   
28              Paul George 22.50000 47.00000 40.90000 89.70000  5.30000   
36             Jimmy Butler 21.40000 50.10000 41.50000 85.60000  5.40000   
6             Jalen Brunson 27.40000 47.70000 40.10000 84.50000  3.60000   
8              Devin Booker 27.20000 49.20000 36.40000 88.10000  4.80000   
15             LeBron James 25.30000 53.00000 40.50000 74.80000  7.30000   
5          Donovan Mitchell 27.40000 46.80000 37.30000 86.20000  5.30000   
33       Karl-Anthony Towns 22.10000 50.60000 42.30000 87.40000  8.40000   
17            Anthony Davis 24.70000 55.30000 28.90000 81.50000 12.60000   
14             Tyrese Maxey 25.80000 44.90000 37.90000 86.70000  3.70000   
9             Stephen Curry 26.50000 44.70000 40.30000 92.30000  4.40000   
18           Damian Lillard 24.40000 42.60000 35.80000 92.30000  4.40000   
24          Lauri Markkanen 23.20000 48.00000 39.90000 89.90000  8.20000   
25          Zion Williamson 22.80000 58.60000 35.70000 70.20000  5.80000   
12          Anthony Edwards 26.20000 46.40000 37.30000 82.70000  5.50000   
34            Pascal Siakam 21.80000 53.50000 34.10000 73.00000  7.00000   
22            DeMar DeRozan 23.30000 47.40000 33.00000 85.90000  4.30000   
37           Alperen Sengun 21.10000 53.70000 29.70000 69.30000  9.30000   
10             De'Aaron Fox 26.50000 46.60000 36.30000 73.00000  4.50000   
11               Trae Young 26.40000 42.60000 37.10000 85.60000  2.70000   
26          Dejounte Murray 22.70000 46.40000 36.60000 80.40000  5.30000   
20             Desmond Bane 23.70000 46.40000 38.10000 87.00000  4.40000   
19            Julius Randle 24.00000 47.20000 31.10000 78.10000  9.20000   
23             Jaylen Brown 23.30000 50.40000 35.40000 69.90000  5.50000   
32          Cade Cunningham 22.30000 44.80000 35.00000 86.60000  4.30000   
27          Anfernee Simons 22.60000 43.00000 38.50000 91.60000  3.60000   
29           Paolo Banchero 22.50000 46.40000 35.90000 71.00000  6.80000   
35               Cam Thomas 21.80000 44.40000 35.70000 84.90000  3.20000   
30               Kyle Kuzma 22.40000 46.70000 33.90000 77.90000  6.50000   
31        Jaren Jackson Jr. 22.30000 44.20000 32.00000 80.50000  5.60000   

        AST     TOV      PF      +/-  desc_preorder  asc_preorder  ranks  \
0   5.70000 3.70000 2.90000  6.90000              1             1      1   
13  9.00000 2.90000 2.50000  8.20000              2             1      2   
3   6.30000 2.20000 2.50000  7.90000              3             3      3   
1   9.80000 4.00000 2.10000  3.50000              7             2      3   
21  3.60000 1.70000 1.40000  5.80000              5             4      4   
4   5.20000 3.30000 1.70000  3.80000              4             8      4   
7   4.90000 2.50000 2.00000  8.80000              6             6      5   
16  5.30000 1.70000 1.90000  3.70000              9             5      5   
2   6.40000 3.40000 2.90000  4.60000             23             4      5   
28  3.60000 2.00000 2.80000  6.30000             10             5      6   
36  4.80000 1.70000 1.20000  3.10000              8             7      6   
6   6.50000 2.40000 1.90000  6.10000             14             9      7   
8   6.90000 2.50000 3.00000  4.60000             14             9   

# PROMETHEE

In [14]:
PROMETHEE = {
    'PTS': {'type': 'gain', 'k': 8, 'q': 4, 'p': 8},
    'FG%': {'type': 'gain', 'k': 4, 'q': 3, 'p': 4},
    '3P%': {'type': 'gain', 'k': 4, 'q': 4, 'p': 5},
    'FT%': {'type': 'gain', 'k': 2, 'q': 3, 'p': 7},
    'REB': {'type': 'gain', 'k': 5, 'q': 2, 'p': 5},
    'AST': {'type': 'gain', 'k': 5, 'q': 1.5, 'p': 3 },
    'TOV': {'type': 'cost', 'k': 3, 'q': 1, 'p': 1.5},
    'PF':  {'type': 'cost', 'k': 1, 'q': 1, 'p': 1.5},
    '+/-': {'type': 'gain', 'k': 2, 'q': 3, 'p': 4},
}

In [15]:
def calculate_marginal_preference(ga, gb, q, p, type_='gain'):
    if type_ == 'gain':
        if ga - gb > p:
            return 1
        elif ga - gb <= q:
            return 0
        else:
            return ((ga-gb)-q)/(p-q)
    elif type_ == 'cost':
        if gb - ga > p:
            return 1
        elif gb - ga <= q:
            return 0
        else:
            return ((gb-ga)-q)/(p-q)
    else:
        raise Exception("wrong criterion type")

## Calculating Comprehensive Preference Index

In [16]:
# Comprehensive Preference Index will be stored inside df_C variable which is a NxN dataframe
df_PI = pd.DataFrame(1.0, index=df.index, columns=df.index)
weights = sum([config['k'] for config in PROMETHEE.values()])

for alt_1_id in df_PI.index:
    for alt_2_id in df_PI.index:
        if alt_1_id == alt_2_id:
            continue

        comprehensive_preference_index = 0
        # calculate marginal preference index for each criterion and add it to the cp index with appropriate weight
        for criterion, config in PROMETHEE.items():
            marginal_preference = calculate_marginal_preference(
                ga=df.loc[alt_1_id, criterion],
                gb=df.loc[alt_2_id, criterion],
                q=config['q'],
                p=config['p'],
                type_=config['type']
            )
            comprehensive_preference_index += config['k'] * marginal_preference
        # normalize the cp index
        comprehensive_preference_index /= weights
        df_PI.loc[alt_1_id, alt_2_id] = comprehensive_preference_index

## Calculating Positive and Negative Flows and PROMETHEE rankings

In [17]:
# calculating flows
df['P_negative_flow'] = df_PI.sum()
df['P_positive_flow'] = df_PI.sum(axis=1)
df['P_balance'] = df['P_positive_flow'] - df['P_negative_flow']

# calculating negative and positive ranking
df['P_ranking_negative'] = df['P_negative_flow'].rank(method='min').astype(int)
df['P_ranking_positive'] = df['P_positive_flow'].rank(ascending=False, method='min').astype(int)

df_ranking_promethee_I = pd.DataFrame(0, index=df.index, columns=df.index)
for alt_1_id in df_ranking_promethee_I.index:
    for alt_2_id in df_ranking_promethee_I.columns:
        if alt_1_id == alt_2_id:
            continue

        alt_1_n = df.loc[alt_1_id, 'P_negative_flow']
        alt_1_p = df.loc[alt_1_id, 'P_positive_flow']
        alt_2_n = df.loc[alt_2_id, 'P_negative_flow']
        alt_2_p = df.loc[alt_2_id, 'P_positive_flow']
        # preference
        if (alt_1_p > alt_2_p and alt_1_n < alt_2_n) or (alt_1_p == alt_2_p and alt_1_n < alt_2_n) or (alt_1_p > alt_2_p and alt_1_n == alt_2_n):
            df_ranking_promethee_I.loc[alt_1_id, alt_2_id] = 1
        elif alt_1_p == alt_2_p and alt_1_n == alt_2_n:
            df_ranking_promethee_I.loc[alt_1_id, alt_2_id] = 1

# ---------------------------------------------------------------------------------------------------- #
# Code used to produce output for graph visualization
            
# relations = []
# for alt_1_id in df_ranking_promethee_I.index:
#     for alt_2_id in df_ranking_promethee_I.columns:
#         if alt_1_id == alt_2_id:
#             continue
#         if df_ranking_promethee_I.loc[alt_1_id, alt_2_id]:
#             relations.append({"alternative_1": alt_1_id, "alternative_2": alt_2_id})

# alternatives = []
# for alt_1_id in df.index:
#     alternatives.append({"id": alt_1_id, "name": df.loc[alt_1_id, "PLAYER"]})

# print(relations)
# print(alternatives)